### ISSUE TITLE
Functional testing triggers NotFoundHttpException "The route is wrongly configured"

### ISSUE THREAD

**Issue by lgugl** [api-platform/core#5923]
**API Platform version(s) affected**: 3.2.0 (and higher)

**Description**  
Two (or more) functional tests in a row, using JWT authentication request, trigger an error if the first test return an exception:
`[error] Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "Unable to find the controller for path "/api/login_check". The route is wrongly configured." at vendor/symfony/http-kernel/HttpKernel.php line 165`

It is "reseted" with a test which does not request authentication. 

**How to reproduce**  
Require lexik/jwt-authentication-bundle, symfony/test-pack, symfony/http-client

Simple JWT configuration (https://symfony.com/bundles/LexikJWTAuthenticationBundle/current/index.html)
(+ with a database and one user: email = username@foo.bar, password = foobar)

Add in config/security.yaml:
```
    access_control:
        - { path: ^/api/foobar/3$, roles: PUBLIC_ACCESS }
        - { path: ^/api/foobar/[0-9]+$, roles: IS_AUTHENTICATED_FULLY }
```

Add in config/packages/api_platform.yaml, under formats:
```
json: ['application/json']
```

Create a DTO in src/ApiResource:
```PHP
<?php
// src/ApiResource/Foobar.php

namespace App\ApiResource;

use ApiPlatform\Metadata\Get;
use App\State\FoobarProvider;

#[Get(
    uriTemplate: '/foobar/{id}',
    provider: FoobarProvider::class
)]

class Foobar
{
}
```

Create the provider:
```PHP
<?php
// src/State/FoobarProvider.php

namespace App\State;

use ApiPlatform\Metadata\Operation;
use ApiPlatform\State\ProviderInterface;
use Symfony\Component\HttpKernel\Exception\HttpException;

class FoobarProvider implements ProviderInterface
{
    public function provide(Operation $operation, array $uriVariables = [], array $context = []): object|array|null
    {
        if ($uriVariables['id'] === '1' || $uriVariables['id'] === '4') {
            // throw new HttpException(400, "there is an issue with this as well");
            return null;
        }

        return [];
    }
}
```

Create a test class:
```PHP
<?php
// tests/FoobarTest.php

namespace App\Tests;

class FoobarTest extends FoobarApiTestCase
{
    public function testOne()
    {
        $response = $this->request('GET', '/api/foobar/1'); // return 404 not found
        $this->assertTrue($response->getStatusCode() >= 300);
    }

    public function testTwo()
    {
        $response = $this->request('GET', '/api/foobar/2'); // triggers the error
        $this->assertResponseIsSuccessful();
    }

    public function testThree()
    {
        $response = $this->publicRequest('GET', '/api/foobar/3'); // it's OK
        $this->assertResponseIsSuccessful();
    }

    public function testFour()
    {
        $response = $this->request('GET', '/api/foobar/4'); // it's OK
        $this->assertResponseIsSuccessful();
    }

    public function testFive()
    {
        $response = $this->request('GET', '/api/foobar/5'); // return 404 not found
        $this->assertTrue($response->getStatusCode() >= 300);
    }

    public function testSix()
    {
        $response = $this->request('GET', '/api/foobar/6'); // triggers the error
        $this->assertResponseIsSuccessful();
    }

}
```

```PHP
<?php
// tests/FoobarApiTestCase.php

namespace App\Tests;

use ApiPlatform\Symfony\Bundle\Test\ApiTestCase;
use Symfony\Contracts\HttpClient\HttpClientInterface;
use Symfony\Contracts\HttpClient\ResponseInterface;

abstract class FoobarApiTestCase extends ApiTestCase
{
    protected HttpClientInterface $client;
    protected string $token;

    protected function setUp(): void
    {
        $this->client = static::createClient();
    }

    protected function getToken(): string
    {
        if (!isset($this->token)) {
            $this->auth();
        }

        return $this->token;
    }

    protected function auth(): void
    {
        $response = $this->client->request('POST', '/api/login_check', [
            'json' => [
                'username' => 'username@foo.bar',
                'password' => 'foobar'
            ]
        ]);

        // $this->assertResponseIsSuccessful();
        $body = json_decode($response->getContent(), true);
        $this->token = $body['token'];
    }

    protected function request(string $method, string $path): ResponseInterface
    {
        $options = [
            'auth_bearer' => $this->getToken(),
            'headers' => [
                'accept' => 'application/json'
            ],
            'json' => []
        ];

        return $this->client->request($method, $path, $options);
    }

    protected function publicRequest(string $method, string $path): ResponseInterface
    {
        $options = [
            'headers' => [],
            'json' => []
        ];

        return $this->client->request($method, $path, $options);
    }
}
```

Then launch the test: `php bin/phpunit`.
==> Failure on testTwo and testSix.
But `php bin/phpunit --filter=testTwo` is green (same with testSix), because no previous request with 400+ response status code.

**Possible Solution**  

**Additional Context**  
Downgrade the API Platform's version to 3.1.21 (or lower), and the error disappears.

**Comment by jeremyrncp** [comment]
Amazing, i don't know where the error came from, but it's surprising !

**Comment by vincenttouzet** [comment]
I've made some digging and found the error but I don´t know how to fix it.

When an error is returned it is transformed into an Error Resource. Then in the [ContentNegotiationProvider](https://github.com/api-platform/core/blob/main/src/State/Provider/ContentNegotiationProvider.php#L47) the format of the request is changed : 

from : 
```
[
  'json'=> ['application/json']
]
```
to 
```
[
  'json'=> ['application/problem+json']
]
```

The `Symfony\Component\HttpFoundation\Request::$formats` is static and therefore it is not changed between the two requests during your tests. In your second test during the login the `Symfony\Component\Security\Http\Authenticator\JsonLoginAuthenticator::supports` checks the content-type of the request but cannot find the type for `application/json` as it was replace in the previous request ...

As a workaround I've added this line in a setUp function in my test : 

```php
(new \Symfony\Component\HttpFoundation\Request())->setFormat('json', ['application/json']);
```
